### ISSUE TITLE
twisted.web leaks at noticeable rate

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11265]
|<img alt="alexm's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| alexm reported|
|-|-|
|Trac ID|trac#589|
|Type|defect|
|Created|2004-04-08 23:37:11Z|

<details><summary>Searchable metadata</summary>

```
trac-id__589 589
type__defect defect
reporter__alexm alexm
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1081467431000000 1081467431000000
changetime__1479164404989061 1479164404989061
version__ 
owner__ 
cc__exarkun cc__spiv cc__itamarst cc__dp cc__PenguinOfDoom cc__hypatia cc__alexm
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="alexm's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|alexm commented|
|-|-|

```
#!html
<pre>
I got a problem with memory leaks in twisted.web. 
 
I have a quite simple test server: 
 
--------------------------------------------------- 
from twisted.application import internet, service 
from twisted.internet import protocol 
from twisted.web import server 
from twisted.web import resource 
# other required imports 
 
#... 
 
class TestResource(resource.Resource): 
    __implements__ = (resource.IResource,) 
    isLeaf = True 
    def __init__(self): 
        resource.Resource.__init__(self) 
    def render(self, request): 
        return 'pong' 
 
root = TestResource() 
 
application = service.Application('Test Server') 
serviceCollection = service.IServiceCollection(application) 
internet.TCPServer(8000,server.Site(root) 
                   ).setServiceParent(serviceCollection) 
--------------------------------- 
Even this basic HTTP server leaks. I ran httperf --hog --server localhost --port 8000 
--num-conn 100000 --ra 1000 --timeout 1 and got leak about 600K per minute. If I change 
resource to smth more useful I got much worse leaks (resulting in 300MB in VmSize according 
to /proc/&lt;pid>/status after a night of work) 
 
I use python-2.3.3 / twisted-1.2.0 on linux-2.4.x and glibc-2.2.6
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|PenguinOfDoom commented|
|-|-|

```
#!html
<pre>
Neither me nor exarkun are able to reproduce this behavior.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
I see a very small leak on Twisted SVN / Linux 2.6.0-test4 / libc6 2.3.2.ds1-11
/ 2.3.3-6.  (On the order of 1Mb per 100000 requests).

I'll have a play around and see if I can determine a cause or even a fix.

Btw, what were the "# other required imports" and "#..." comments about?   The
example works just fine as-is.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Oh, and my machine is a 750Mhz Athlon.  I mention this in case this is timing
related...
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Ok, more information:

The leak I see is more like 1Mb per 10000 requests.  The leak doesn't seem to be
Python objects, though -- len(gc.get_objects()) isn't growing (it fluctuates,
but never rises above 19000 that I can see).  There are also no unreachable
cycles accumulating in gc.garbage.

So it looks like I'm seeing a problem at the C level.  It might be a bug in one
of the standard extension modules like socket (although I doubt it), or it might
just be plain old memory fragmentation due to squillions of object allocations
and deallocations, although if that were the case I'd expect the memory use to
plateau eventually, and this seems to be rising constantly.

Suggestions on how to pin this down further are welcome.  At this stage the best
I can think of is probably valgrind, but given how slow this leak is, that's
going to take me a while to run, let alone to wade through the results...
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="alexm's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|alexm commented|
|-|-|

```
#!html
<pre>
I wonder if this BTS can be run from e-mail? I sent two mail messages but it
seems they're simply ignored
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="alexm's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|alexm commented|
|-|-|

```
#!html
<pre>
Well, can it be the issue mentioned here (or similar):

http://mail.python.org/pipermail/python-list/2003-August/177299.html

?

But that issue marked as resolved already...
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
are we still leaking memory?  is this related to the quotient issue we were seeing recently?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
I cannot reproduce with given test program nor any other i've run. Ran a couple hundred thousand hits 
through it and memory stayed constant.

Python 2.3 (#1, Sep 13 2003, 00:49:11) 
[GCC 3.3 20030304 (Apple Computer, Inc. build 1495)] on darwin
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="k3mper's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|k3mper commented|
|-|-|

```
#!html
<pre>
Maybe I have same memory leak problem. Running that test server send by alexm
memory usage increases by 400-600KB every 10000 requests. Object count stays
~16000. 

ActivePython 2.3.2 Build 232 on win32 + Twisted-1.3.0
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
I've just checked in revision r10970 which fixes a memory leak in the _c_urlarg
C module.  Can people who were seeing this leak please grab the latest SVN and
test and report back to this bug, please?

Note that if you didn't build the C extensions (i.e. setup.py build_ext -i), you
wouldn't have triggered this bug -- which may explain why some people didn't see it.

Thanks!
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
I should mention that this bug is now fixed for me (memory usage stays constant
while hammering the test server in this bug), so unless someone speaks up, I'll
close this bug in the next few days.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Do we want c_url_arg at all? any reason we shouldn't just drop it?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
It was originally written because URL decoding in Python was a significant hit
to performance -- this module gained us several percent in simple benchmarks, iirc.

It's also a dead-simple module, and I've previously reviewed it trying to crash
it, and couldn't.  There are tests for edge-cases that it gracefully handles in
twisted.test.test_http.QueryArgumentsTestCase.

So I'm personally happy to keep it, because I consider it rock-solid, and it is
considerably faster than urllib.unquote:

$ python2.3 /usr/lib/python2.3/timeit.py -s "from urllib import unquote"
"unquote('%ab%cd%xyjfdsklasfjks%%sdfk%12%f0fdjk')"
10000 loops, best of 3: 104 usec per loop

$ python2.3 /usr/lib/python2.3/timeit.py -s "from twisted.protocols._c_urlarg
import unquote" "unquote('%ab%cd%xyjfdsklasfjks%%sdfk%12%f0fdjk')"
100000 loops, best of 3: 3.57 usec per loop
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="spiv's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|spiv commented|
|-|-|

```
#!html
<pre>
Ok, no-one has reported a leak with current SVN, so I'm closing this issue.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_17"></a>|@rodrigc commented|
|-|-|

In changeset 9547189afbd4d0c47c609c9031f2322f39de8c5d
```
#!CommitTicketReference repository="" revision="9547189afbd4d0c47c609c9031f2322f39de8c5d"
Merge pull request #589 from twisted/8791-rodrigc-conch

Author: rodrigc
Reviewer: acabhishek942
Fixes: #8791

Fix conch and cftp under Python 3
```
